File: apps/web/src/unified/inspectorPresentation.ts
export type InspectorPart =
  | { type: "text" | "number"; value: string }
  | { type: "date" | "datetime"; value: string }
  | { type: "money"; value: string; currency: string; precision?: number };

export function renderInspectorValue(
  value: unknown,
  parts: InspectorPart[] | undefined,
  format: {
    number: (value: string) => string;
    money: (value: string, currency: string, precision?: number) => string;
    date?: (value: string) => string;
    dateTime?: (value: string) => string;
  },
): string {
  const readableText = (text: string) =>
    format.dateTime
      ? text.replace(
          /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})/g,
          (value) => format.dateTime!(value),
        )
      : text;
  if (!parts)
    return typeof value === "object" && value !== null
      ? JSON.stringify(value)
      : String(value ?? "—");
  return parts
    .map((part) =>
      part.type === "money"
        ? format.money(part.value, part.currency, part.precision)
        : part.type === "date"
          ? (format.date?.(part.value) ?? part.value)
          : part.type === "datetime"
            ? (format.dateTime?.(part.value) ?? part.value)
            : part.type === "number"
              ? format.number(part.value)
              : readableText(part.value),
    )
    .join("");
}

File: packages/reality-core/src/reality/services/inspector_presentation.py
"""Optional locale-aware presentation parts; original Inspector values stay compatible."""

from datetime import date, datetime
from decimal import Decimal
from typing import Any


class DisplayText(str):
    parts: list[dict[str, Any]]

    def __new__(cls, value: str, parts: list[dict[str, Any]]):
        result = super().__new__(cls, value)
        result.parts = parts
        return result


def display_parts(value: Any) -> list[dict[str, Any]] | None:
    if isinstance(value, DisplayText):
        return value.parts
    if isinstance(value, datetime):
        return [{"type": "datetime", "value": value.isoformat()}]
    if isinstance(value, date):
        return [{"type": "date", "value": value.isoformat()}]
    if isinstance(value, (Decimal, int)) and not isinstance(value, bool):
        return [{"type": "number", "value": str(value)}]
    return None


def display_text(*values: Any) -> DisplayText:
    parts = []
    for value in values:
        parts.extend(display_parts(value) or [{"type": "text", "value": str(value)}])
    return DisplayText("".join(str(value) for value in values), parts)


def money(
    value: Any, currency: str | None, *, precision: int | None = None
) -> DisplayText:
    if not currency:
        return display_text(value)
    return DisplayText(
        f"{currency} {value}",
        [
            {
                "type": "money",
                "value": str(value),
                "currency": currency,
                **({"precision": precision} if precision is not None else {}),
            }
        ],
    )
